lib/portage/package/ebuild/doebuild.py

```diff
@@ -2560,17 +2552,51 @@ def _post_src_install_write_metadata(settings):
                 except OSError:
                     pass
                 continue
-            with open(
-                _unicode_encode(
-                    os.path.join(build_info_dir, k),
-                    encoding=_encodings["fs"],
-                    errors="strict",
-                ),
-                mode="w",
-                encoding=_encodings["repo.content"],
+
+            metadata_buffer[k] = v
+
+    # We could skip this for non-binpkgs but there doesn't seem to be much
+    # value in that, as users shouldn't downgrade libc anyway.
+    #
+    # TODO: We could skip the injection for packages installing no ELF files
+    # by going off the absence of NEEDED, but that runs after this function.
+    # Maybe we should just move the scanelf calls into Python and/or use pyelftools
+    # instead which would let us coalesce the calls too.
+    injected_libc_depstring = []
+    for atom in expand_new_virt(
+        QueryCommand.get_db()[settings["EROOT"]]["vartree"].dbapi,
+        portage.const.LIBC_PACKAGE_ATOM,
+    ):
+        # This will give us something like sys-libs/glibc:2.2, but we want to know
+        # what installed atom actually satifies that.
+        try:
+            libc_realized_atom = QueryCommand.get_db()[settings["EROOT"]][
+                "vartree"
+            ].dbapi.match(atom)[0]
+        except IndexError:
+            continue
+
+        injected_libc_depstring.append(f">={libc_realized_atom}")
+
+    # For RDEPEND, we want an implicit dependency on >=${PROVIDER_OF_LIBC}
+    # to avoid runtime breakage when merging binpkgs, see bug #753500.
+    if injected_libc_depstring and settings.configdict["pkg"].get("RDEPEND"):
```

Comment:
I'm going to see about only doing this for binpkgs to avoid the changed-deps issue Zac mentions in the bug.